Adding smotenc (for categorical variables)

RobertGregg opened the pull request:
This is first pull request and I'm also a new-comer to R development, so bare with me. 

I wanted to add a smote algorithm that could incorporate numeric and categorical variables, similar to [scikit](https://imbalanced-learn.org/dev/references/generated/imblearn.over_sampling.SMOTENC.html). I mostly patterned matched the code from `step_smote()` and `smote_impl()`.

There are two upgrades I added to create `step_smotenc()`

1. Replace the `RANN::nn2()` nearest neighbor search with `gower::gower_topn()`
2. Synthetically create new categorical variable instances with the most common category determined from the nearest neighbors

The [original SMOTE paper]( https://doi.org/10.1613/jair.953) has a very convoluted method of including a calculated median value if the categories of the nominal feature are the same. We can remove that complexity by just using a distance metric designed with mixed data in mind (I chose to use the [gower distance](https://jamesmccaffrey.wordpress.com/2020/04/21/example-of-calculating-the-gower-distance/) because it seems like the most common metric for this application and there was already an R package implementation). The second upgrade follows the paper's recommendation. I also removed the "No categorical variables" check for obvious reasons. 

Side note: I also added two functions into `misc.R` called `Mode()` and `is.not.numeric()`. (Why R doesn't have a statistical mode function is baffling to me...) 

I updated the files using the styler package but ran into issues when testing. When I run `step_smotenc()` in global scope the step runs as expected. However when I run `test_check("themis")`, I get an error:

```r
Error in `UseMethod("prep")`: no applicable method for 'prep' applied to an object of class "c('step_smotenc', 'step')"
``` 

I've defined the `prep()` method in `step_smotenc.R` but it can't find the function for some reason. 

I also don't know how to use roxygen2, maybe that caused an error somewhere? I figured I would focus on documenting after it was working, but maybe that was a mistake.

Hopefully with a little help this could get up and running. Let me know what you think.

EmilHvitfeldt commented:
Hello @RobertGregg thank you so much for taking on this step!

The error you are getting is caused because we need to tell the package what functions to export. If you run `devtools::document()` it should update the `NAMESPACE` file and a bunch of `.Rd` files.  Run `devtools::document()`, commit those changes. Hopefully, that will allow you to finish up the testing. Ideally, you would want to be able to run `devtools::check()` with no errors or warnings, if you get stuck let me know

EmilHvitfeldt commented on R/step_smotenc.R:
```diff
@@ -0,0 +1,210 @@
+print.step_smotenc <-
+  function(x, width = max(20, options()$width - 26), ...) {
+    cat("SMOTE based on ", sep = "")
+    printer(x$column, x$terms, x$trained, width = width)
```
This is a very recent change, you change this to follow this format:
https://github.com/tidymodels/themis/blob/8698526bda3381b53ccf120057cecc8d173bb5ab/R/step_smote.R#L193-L199

Also make sure that the title for this step says SMOTENC instead of SMOTE

EmilHvitfeldt reviewed (commented):
Looks, good so far. Once you do the above things I mentioned and get it working I'll go over al do another review 😃

RobertGregg commented on R/smotenc_impl.R:
```diff
@@ -0,0 +1,174 @@
+
+
+# Splits data and appends new minority instances
+smote_impl <- function(df, var, cat_vars, k, over_ratio) {
```
Updated `smote_impl` -> `smotenc_impl`

RobertGregg commented on R/smotenc_impl.R:
```diff
@@ -0,0 +1,174 @@
+}
+
+# Uses nearest-neighbors and interpolation to generate new instances
+smote_data <- function(data, k, n_samples, smote_ids = seq_len(nrow(data))) {
```
Updated `smote_data` -> `smotenc_data`

RobertGregg commented:
Glad I can help! Okay, things are looking a lot better after I used `devtools::document()` and `devtools::check()` in the new commit. 

I updated the example in `step_smotenc` to include a categorical variable and changed the variable names as you mentioned (thank you for catching that). 

`devtools::check()` gives a warning about not importing `gower` which makes sense but I'm not sure how to add the dependency. It also suggests the `covr` package, but I'm not sure why.

There are 3 test failures. 

1. Failure (test-smotenc.R:84:3): errors if character are present
2. Failure (test-smotenc.R:285:3): empty printing
3. Failure (test-smotenc.R:289:3): empty printing

The first test should be replaced with something checking if `step_smotenc` can handle character/factor variables (maybe using the example I created mentioned above?). The last two mention running `snapshot_accept('smotenc')` if the change was deliberate, but I didn't want to do that without understanding what it meant.

EmilHvitfeldt commented:
It looks good! You can add `gower` as a dependency by using `usethis::use_package("gower")`. for the tests, the "errors if character are present" tests should ideally be deleted since this step works with characters, and should thus not error. For the other two, you can run `testthat::snapshot_accept('smotenc')` after running the tests to accept the changes.

After you do that I'll take over a do a couple of small changes. Thank you so much for your effort!

RobertGregg commented:
Awesome. The `gower` package was added and I replaced the character failure test to check if it could run successfully with a character input. I also ran`testthat::snapshot_accept('smotenc')` but the failed test persisted saying there were "No snapshots to update".    

```r
── Failure (test-smotenc.R:289:3): empty printing ──────────────────────────────
Snapshot of code has changed:
`old[13:16]`: "  " "  Operations:" "  " "  SMOTE based on <none> [trained]"  
`new[13:16]`: "  " "  Operations:" "  " "  SMOTENC based on <none> [trained]"
```
Perhaps just a small hiccup in the printing function? 

Overall this was a really smooth process, looking forward to seeing the step in the next release 😁

EmilHvitfeldt commented:
Thank you for all the work! I did a little bit of cleanup, and a slight speedup by performing the calculations in matrices instead of data.frames.
